### ISSUE TITLE
scylla-sstable: add a `download` command, which downloads an sstable located in S3, to local storate

### ISSUE THREAD

**Issue by denesb** [scylladb/scylladb#20537]
Allow downloading sstables located in S3 to local storage.

Proposed syntax for download:
```
scylla sstable download [--endpoint=$endpoint] [--concurrency=$concurrency] --output-dir=$out_dir $sstable_url
```

See https://github.com/scylladb/scylladb/issues/20532 for the meaning of common object-storage command-line parameters, as well as how S3 configuration is obtained.
* `$out_dir` - the local path on disk where the downloaded files will be placed. Has the same semantics as `--output-dir` from [scylla-sstable write](https://opensource.docs.scylladb.com/stable/operating-scylla/admin-tools/scylla-sstable.html#write).
* `$concurrency` (optional, default=1) - how many files to download concurrently.

**Comment by tchaikov** [comment]
> The last point contains some unvalidated assumptions from my end, w.r.t. how sstables on S3 are laid out and how they can be accessed. @tchaikov @xemul please validate my assumptions.

you mean the `$file_name_pattern` part? i think it makes sense. we encode the filename as the last component of the corresponding object's key . so, along with `endpoint`, `bucket` and `prefix`, they are actually flexible enough to represent all object keys, even if the keys of object not storing sstables. FWIW, [S3 API](ListObjectsV2) does not provide the filter option, so we will need to filter the listed object keys on the client (i.e., `scylla sstable`) side.

**Comment by denesb** [comment]
> > The last point contains some unvalidated assumptions from my end, w.r.t. how sstables on S3 are laid out and how they can be accessed. @tchaikov @xemul please validate my assumptions.
> 
> you mean the `$file_name_pattern` part? i think it makes sense. we encode the filename as the last component of the corresponding object's key . so, along with `endpoint`, `bucket` and `prefix`, they are actually flexible enough to represent all object keys, even if the keys of object not storing sstables. FWIW, [S3 API](ListObjectsV2) does not provide the filter option, so we will need to filter the listed object keys on the client (i.e., `scylla sstable`) side.

Yes, that is what I mean. I also saw that you can encode paths as: `s3://bucket/prefix/file_name`, so it is quite flexible. If encoded like this, one only has to provide the endpoint. Am I correct?

**Comment by tchaikov** [comment]
> I also saw that you can encode paths as: `s3://bucket/prefix/file_name`, 

we are not using this representation for object key (yet). but it's indeed a nice way to represent a key or a key glob pattern in `scylla sstable`, so one can use 
```shell
scylla sstable download --endpoint=$endpoint $bucket/$prefix/*.TOC.txt`
```

because endpoint is hardwired to a given backend, and currently, only S3 is supported. so the `s3://` schema could be redundant, and probably misleading. as it would be always ignored.

> so it is quite flexible. If encoded like this, one only has to provide the endpoint. Am I correct?

yes, kind of. as explained above.

**Comment by denesb** [comment]
> because endpoint is hardwired to a given backend, and currently, only S3 is supported. so the `s3://` schema could be redundant, and probably misleading. as it would be always ignored.



> > I also saw that you can encode paths as: `s3://bucket/prefix/file_name`,
> 
> we are not using this representation for object key (yet). but it's indeed a nice way to represent a key or a key glob pattern in `scylla sstable`, so one can use
> 
> ```shell
> scylla sstable download --endpoint=$endpoint $bucket/$prefix/*.TOC.txt`
> ```
> 
> because endpoint is hardwired to a given backend, and currently, only S3 is supported. so the `s3://` schema could be redundant, and probably misleading. as it would be always ignored.
> 

RIght, so `s3://` is more relevant for https://github.com/scylladb/scylladb/issues/20535, as a way to tell apart local files from remote ones, allowing us to mix them (although this arguably has low value). Maybe we can use the `s3://` just for consistency.
Actually, since we can control how we interpret URL:s, we can change it to be `s3://$endpoint/$bucket/$prefix/$path`, and thus make it self-contained. Does this go against some already existing S3 convention?

**Comment by tchaikov** [comment]
> RIght, so `s3://` is more relevant for #20535, as a way to tell apart local files from remote ones, allowing us to mix them (although this arguably has low value). Maybe we can use the `s3://` just for consistency. 

i see. probably an alternative is use something like `$endpoint:$bucket/$prefix/$path` or `$endpoint:/$bucket/$prefix/$path`. but i admit that neither of them looks like an URL =( . but i am actually trying to mimic 

```shell
scp /tmp/my.sav ps5:/var/lib/
```

> Actually, since we can control how we interpret URL:s, we can change it to be `s3://$endpoint/$bucket/$prefix/$path`, and thus make it self-contained. 
> Does this go against some already existing S3 convention?

no, AFAICT. on the contrary, i think it's quite a popular way to encode an S3 object's URI. at least AWS's tools use it, see https://docs.aws.amazon.com/cli/latest/reference/s3/#path-argument-type and https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_S3StorageConfig.html .

**Comment by denesb** [comment]
> no, AFAICT. on the contrary, i think it's quite a popular way to encode an S3 object's URI. at least AWS's tools use it, see https://docs.aws.amazon.com/cli/latest/reference/s3/#path-argument-type and https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_S3StorageConfig.html .

In both the above linked examples, I don't see the endpoint included in the URI, only the bucket and optionally the prefix.
So I think the `scp`-style syntax might work better. We can accept both `scp` style syntax, or `s3://` syntax, with either implicit `endpoint` (take it from config, but only if config has single endpoint) or explicit one (provided via `--endpoint`).

**Comment by tchaikov** [comment]
> In both the above linked examples, I don't see the endpoint included in the URI, only the bucket and optionally the prefix.

ahh, i missed it. yeah, endpoint is configured with configuration file, environment variable or `--endpoint-url`. see https://docs.aws.amazon.com/cli/v1/userguide/cli-configure-files.html#cli-config-endpoint_url .

> So I think the `scp`-style syntax might work better. We can accept both `scp` style syntax, or `s3://` syntax, with either implicit `endpoint` (take it from config, but only if config has single endpoint) or explicit one (provided via `--endpoint`).

agreed. this way, user would be facing a probably familiar user interface, and it's flexible enough to support other backends in future.

**Comment by denesb** [comment]
Ok, I will update the description based on this.

**Comment by denesb** [comment]
I moved the common S3-related syntax proposal to https://github.com/scylladb/scylladb/issues/20532.

**Comment by denesb** [comment]
Closing, it looks like downloading sstables is simple enough with awscli.
